lib/semantics/check-allocate.cc

```diff
@@ -0,0 +1,411 @@
+  if (const DerivedTypeSpec * derivedType1{type1.AsDerived()}) {
+    if (type1.category() == DeclTypeSpec::Category::TypeDerived) {
+      return &derivedType1->typeSymbol() == &derivedType2.typeSymbol();
+    } else if (type1.category() == DeclTypeSpec::Category::ClassDerived) {
+      for (const DerivedTypeSpec *parent{&derivedType2}; parent != nullptr;
+           parent = parent->typeSymbol().GetParentTypeSpec()) {
+        if (&derivedType1->typeSymbol() == &parent->typeSymbol()) {
+          return true;
+        }
+      }
+    }
+  }
+  return false;
+}
+
+static bool IsTypeCompatible(
+    const DeclTypeSpec &type1, const DeclTypeSpec &type2) {
+  if (type1.category() == DeclTypeSpec::Category::ClassStar) {
+    // TypeStar does not make sens in allocate context because assumed type
+    // cannot be allocatable (C709)
+    return true;
+  }
+  if (const IntrinsicTypeSpec * intrinsicType2{type2.AsIntrinsic()}) {
+    if (const IntrinsicTypeSpec * intrinsicType1{type1.AsIntrinsic()}) {
+      return intrinsicType1->category() == intrinsicType2->category();
+    } else {
+      return false;
+    }
+  } else if (const DerivedTypeSpec * derivedType2{type2.AsDerived()}) {
+    return IsTypeCompatible(type1, *derivedType2);
+  }
+  return false;
+}
+
+static bool IsTypeCompatible(
+    const DeclTypeSpec &type1, const evaluate::DynamicType &type2) {
+  if (type1.category() == DeclTypeSpec::Category::ClassStar) {
+    // TypeStar does not make sens in allocate context because assumed type
+    // cannot be allocatable (C709)
+    return true;
+  }
+  if (type2.category != evaluate::TypeCategory::Derived) {
+    if (const IntrinsicTypeSpec * intrinsicType1{type1.AsIntrinsic()}) {
+      return intrinsicType1->category() == type2.category;
+    } else {
+      return false;
+    }
+  } else {
+    CHECK(type2.derived);
+    return IsTypeCompatible(type1, *type2.derived);
+  }
+  return false;
+}
+
+// Note: Check assumes  type1 is compatible with type2. type2 may have more type
+// parameters than type1 but if a type2 type parameter is assumed, then this
+// check enforce that type1 has it. type1 can be unlimited polymorphic, but not
+// type2.
+static bool HaveSameAssumedTypeParameters(
+    const DeclTypeSpec &type1, const DeclTypeSpec &type2) {
+  if (type2.category() == DeclTypeSpec::Category::Character) {
+    bool type2LengthIsAssumed{type2.characterTypeSpec().length().isAssumed()};
+    if (type1.category() == DeclTypeSpec::Category::Character) {
+      return type1.characterTypeSpec().length().isAssumed() ==
+          type2LengthIsAssumed;
+    }
+    // It is possible to reach this if type1 is unlimited polymorphic
+    return !type2LengthIsAssumed;
+  } else if (const DerivedTypeSpec * derivedType2{type2.AsDerived()}) {
+    int type2AssumedParametersCount{0};
+    int type1AssumedParametersCount{0};
+    for (auto it{derivedType2->parameters().begin()};
+         it != derivedType2->parameters().end(); ++it) {
```

Comment:
Consider using a range-based for loop here and below:
`for (auto pair : derivedType2->parameters())`

Reply:
Applied